src/generic_containers.rs

```diff
@@ -0,0 +1,114 @@
+//! **Private** generalizations of containers:
+//! - `Map`: `BTreeMap` and `HashMap` (any hasher).
+
+#![cfg(feature = "use_alloc")]
+
+use alloc::collections::BTreeMap;
+use alloc::vec::Vec;
+#[cfg(feature = "use_std")]
+use core::hash::{BuildHasher, Hash};
+#[cfg(feature = "use_std")]
+use std::collections::HashMap;
+
+pub trait Map {
+    type Key;
+    type Value;
+    fn is_empty(&self) -> bool;
+    fn clear(&mut self);
+    fn insert(&mut self, key: Self::Key, value: Self::Value) -> Option<Self::Value>;
+    fn remove(&mut self, key: &Self::Key) -> Option<Self::Value>;
+    fn entry_or_default(&mut self, key: Self::Key) -> &mut Self::Value
+    where
+        Self::Value: Default;
+}
+
+impl<K, V> Map for BTreeMap<K, V>
+where
+    K: Ord,
+{
+    type Key = K;
+    type Value = V;
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+    fn clear(&mut self) {
+        self.clear()
+    }
+    fn insert(&mut self, key: K, value: V) -> Option<V> {
+        self.insert(key, value)
+    }
+    fn remove(&mut self, key: &K) -> Option<V> {
+        self.remove(key)
+    }
+    fn entry_or_default(&mut self, key: K) -> &mut V
+    where
+        V: Default,
+    {
+        self.entry(key).or_default()
+    }
+}
+
+#[cfg(feature = "use_std")]
+impl<K, V, S> Map for HashMap<K, V, S>
+where
+    K: Eq + Hash,
+    S: BuildHasher,
+{
+    type Key = K;
+    type Value = V;
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+    fn clear(&mut self) {
+        self.clear()
+    }
+    fn insert(&mut self, key: K, value: V) -> Option<V> {
+        self.insert(key, value)
+    }
+    fn remove(&mut self, key: &K) -> Option<V> {
+        self.remove(key)
+    }
+    fn entry_or_default(&mut self, key: K) -> &mut V
+    where
+        V: Default,
+    {
+        self.entry(key).or_default()
+    }
+}
+
+impl<K, V> Map for Vec<(K, V)>
+where
+    K: PartialEq,
```

Comment:
I chose vectors to be unsorted to have a smaller bound (`PartialEq` instead of `PartialOrd` for sorted ones). So implement `Map` for `Vec<(K, V)>` is possible but _do we keep it?!_

(if kept: add tests, update docs, update doc of "generic_containers" module).

Reply:
I think the point of using a `Vec<(K, V)>` is that it is faster when you have so few items that testing all of them is faster than doing something smarter with sorted or hashed data structures, so for me it makes sense to keep it unsorted.

If there's something I would change is to require `Eq` instead of `PartialEq` to catch possible mistakes with e.g. floating point numbers and to be more in line with the other map types. A counterpoint to this is that `PartialEq` is more flexible and a `Vec<(K, V)>` wouldn't break as hard as a `HashMap`/`BTreeMap` is the `Eq` contract wasn't respected.

Reply:
I `grep`ped `Eq` to see our usage of it and it's only used at the same we use `Hash` (so `HashMap/HashSet`).
However, we could use `Eq` (I like the consistency with other maps) and eventually relax the bound to `PartialEq` if someone comes with a really good argument.